idb/postgres_migrations.go

```diff
@@ -97,6 +103,136 @@ func m2apps(db *PostgresIndexerDb, state *MigrationState) error {
 	return sqlMigration(db, state, sqlLines)
 }
 
+func m3acfgFix(db *PostgresIndexerDb, state *MigrationState) error {
+	go m3acfgFixAsync(db, state)
+	return asyncMigration
+}
+
+func m3acfgFixAsync(db *PostgresIndexerDb, state *MigrationState) {
+	log.Print("asset config fix migration starting")
+	rows, err := db.db.Query(`SELECT index FROM asset WHERE index >= $1 ORDER BY 1`, state.NextAssetId)
+	if err != nil {
+		log.Printf("acfg fix err getting assetids, %v", err)
+		return
+	}
+	assetIds := make([]int64, 0, 1000)
+	for rows.Next() {
+		var aid int64
+		err = rows.Scan(&aid)
+		if err != nil {
+			log.Printf("acfg fix err getting assetid row, %v", err)
+			rows.Close()
+			return
+		}
+		assetIds = append(assetIds, aid)
+	}
+	rows.Close()
+	for true {
+		nexti, err := m3acfgFixAsyncInner(db, state, assetIds)
+		if err != nil {
+			log.Printf("acfg fix chunk, %v", err)
+			return
+		}
+		if nexti < 0 {
+			break
+		}
+		assetIds = assetIds[nexti:]
+	}
+	log.Print("acfg fix migration finished")
+	err = db.runMigrationState(state)
+	if err != nil {
+		log.Print(err)
+	}
+}
+
+// do a transactional batch of asset fixes
+// updates asset rows and metastate
+func m3acfgFixAsyncInner(db *PostgresIndexerDb, state *MigrationState, assetIds []int64) (next int, err error) {
+	lastlog := time.Now()
+	tx, err := db.db.Begin()
+	if err != nil {
+		log.Printf("acfg fix tx begin, %v", err)
+		return -1, err
+	}
+	defer tx.Rollback() // ignored if .Commit() first
+	setacfg, err := tx.Prepare(`INSERT INTO asset (index, creator_addr, params) VALUES ($1, $2, $3) ON CONFLICT (index) DO UPDATE SET params = EXCLUDED.params`)
+	if err != nil {
+		log.Printf("acfg fix prepare set asset, %v", err)
+		return
+	}
+	defer setacfg.Close()
+	for i, aid := range assetIds {
+		now := time.Now()
+		if now.Sub(lastlog) > (5 * time.Second) {
+			log.Printf("acfg fix next=%d", aid)
+			state.NextAssetId = aid
+			migrationStateJson := json.Encode(state)
+			_, err = tx.Exec(setMetastateUpsert, migrationMetastateKey, migrationStateJson)
+			if err != nil {
+				log.Printf("acfg fix migration %d meta err: %v", state.NextMigration, err)
+				return -1, err
+			}
+			err = tx.Commit()
+			if err != nil {
+				log.Printf("acfg fix migration %d commit err: %v", state.NextMigration, err)
+				return -1, err
+			}
+			return i, nil
+		}
+		txrows := db.txTransactions(tx, TransactionFilter{TypeEnum: 3, AssetId: uint64(aid)})
+		prevRound := uint64(0)
+		first := true
+		var params types.AssetParams
+		var creator types.Address
+		for txrow := range txrows {
+			if txrow.Round < prevRound {
+				log.Printf("acfg rows out of order %d < %d", txrow.Round, prevRound)
```

Comment:
Should this be an error? Wouldn't a return here cause the remaining assets to be skipped?

Reply:
It looks like there are a couple ways that this could fail. I think that we need to make sure Indexer is halted with an error rather than letting it be brought up in an inconsistent state.

Reply:
Approving despite this because (a) I trust that postgres can order it's results and (b) algod shouldn't accept invalid transactions. But maybe one of the queries could get dropped and cause a failure as well, so we probably need a way to return errors instead of silently failing to complete a migration.

Reply:
return here causes the migration to stop uncompleted. the migration will run again the next time indexer starts (possibly after a binary upgrade with a bug fix)